File: assign8/dbscan.py
import math

def distance(a, b):
    return math.sqrt(sum((x - y) ** 2 for x, y in zip(a, b)))

def centroid(cluster):
    dims = len(cluster[0][1]) 
    mean = [0.0] * dims
    for _, p in cluster:
        for i in range(dims):
            mean[i] += p[i]
    mean = [x / len(cluster) for x in mean]
    return mean

def read_csv(filename):
    points = []
    with open(filename, "r") as f:
        header_skipped = False
        for line in f:
            line = line.strip()
            if not line:
                continue
            vals = line.split(",")
            if not header_skipped and any("id" in v.lower() for v in vals):
                header_skipped = True
                continue
            try:
                name = vals[0]
                coords = [float(v) for v in vals[1:]]
                points.append((name, coords))
            except:
                continue
    return points

def compute_distance_matrix(points):
    n = len(points)
    matrix = [[0.0]*n for _ in range(n)]
    names = [p[0] for p in points]

    for i in range(n):
        for j in range(n):
            if i == j:
                matrix[i][j] = 0.0
            elif i < j:
                d = distance(points[i][1], points[j][1])
                matrix[i][j] = d
                matrix[j][i] = d

    print("\nDistance Matrix:")
    print("           " + " ".join(f"{name:10}" for name in names))  # header row
    for i, row in enumerate(matrix):
        print(f"{names[i]:10} " + " ".join(f"{x:<10.2f}" for x in row))
    return matrix

def region_query(points, point, eps):
    neighbors = []
    for idx, (_, p) in enumerate(points):
        if distance(p, point[1]) <= eps:
            neighbors.append(idx)
    return neighbors

def expand_cluster(points, clusters, cluster_id, point_idx, neighbors, eps, min_pts):
    clusters[point_idx] = cluster_id
    queue = list(neighbors)
    while queue:
        q_idx = queue.pop(0)
        if clusters[q_idx] == 0:  
            clusters[q_idx] = cluster_id
            q_neighbors = region_query(points, points[q_idx], eps)
            if len(q_neighbors) >= min_pts:
                queue.extend(q_neighbors)
        if clusters[q_idx] == -1:  
            clusters[q_idx] = cluster_id

def dbscan(points, eps, min_pts):
    n = len(points)
    clusters = [0] * n 
    cluster_id = 0
    for i in range(n):
        if clusters[i] != 0:
            continue
        neighbors = region_query(points, points[i], eps)
        if len(neighbors) < min_pts:
            clusters[i] = -1
        else:
            cluster_id += 1
            expand_cluster(points, clusters, cluster_id, i, neighbors, eps, min_pts)

    cluster_map = {}
    for idx, c_id in enumerate(clusters):
        if c_id > 0:
            if c_id not in cluster_map:
                cluster_map[c_id] = []
            cluster_map[c_id].append(points[idx])
    return list(cluster_map.values()), clusters


filename = input("Enter the CSV filename: ")
points = read_csv(filename)

compute_distance_matrix(points)

eps = float(input("\nEnter epsilon: "))
min_pts = int(input("Enter minPts: "))

clusters, cluster_assignments = dbscan(points, eps, min_pts)

total_points = len(points)
print("\nResults:")
for i, cluster in enumerate(clusters, start=1):
    percent = len(cluster) * 100.0 / total_points
    names = [p[0] for p in cluster]
    print(f"Cluster {i}: {len(cluster)} points ({percent:.2f}%) → {names}")

noise_points = [points[i][0] for i, c in enumerate(cluster_assignments) if c == -1]
if noise_points:
    noise_count = len(noise_points)
    noise_percent = noise_count * 100.0 / total_points
    print(f"Noise points: {noise_count} ({noise_percent:.2f}%) → {noise_points}")

print("\nCentroids:")
for i, cluster in enumerate(clusters, start=1):
    c = centroid(cluster)
    print(f"Centroid {i}: [{', '.join(f'{x:.2f}' for x in c)}]")

File: assign8/DBSCAN.java
import java.io.*;
import java.util.*;
import java.lang.Math;

class Point {
    String name;
    double[] coords;
    Point(String name, double[] coords) {
        this.name = name;
        this.coords = coords;
    }
}

public class DBSCAN {
    public static double distance(double[] a, double[] b) {
        double sum = 0.0;
        for (int i = 0; i < a.length; i++)
            sum += Math.pow(a[i] - b[i], 2);
        return Math.sqrt(sum);
    }

    public static List<Point> readCSV(String filename) throws IOException {
        List<Point> points = new ArrayList<>();
        BufferedReader br = new BufferedReader(new FileReader(filename));
        String line;
        boolean headerSkipped = false;

        while ((line = br.readLine()) != null) {
            if (line.trim().isEmpty()) continue;
            String[] vals = line.split(",");
            if (!headerSkipped && vals[0].toLowerCase().contains("point")) {
                headerSkipped = true;
                continue;
            }
            try {
                String name = vals[0].trim();
                double[] coords = new double[vals.length - 1];
                for (int i = 1; i < vals.length; i++)
                    coords[i - 1] = Double.parseDouble(vals[i]);
                points.add(new Point(name, coords));
            } catch (Exception e) {
                continue;
            }
        }
        br.close();
        return points;
    }

    public static double[][] computeDistanceMatrix(List<Point> points) {
        int n = points.size();
        double[][] matrix = new double[n][n];

        System.out.print("\nDistance Matrix:\n          ");
        for (Point p : points)
            System.out.printf("%-10s", p.name);
        System.out.println();

        for (int i = 0; i < n; i++) {
            System.out.printf("%-10s", points.get(i).name);
            for (int j = 0; j < n; j++) {
                if (i == j) matrix[i][j] = 0.0;
                else if (i < j) matrix[i][j] = matrix[j][i] = distance(points.get(i).coords, points.get(j).coords);
                System.out.printf("%-10.2f", matrix[i][j]);
            }
            System.out.println();
        }
        return matrix;
    }

    public static List<Integer> regionQuery(List<Point> points, Point point, double eps) {
        List<Integer> neighbors = new ArrayList<>();
        for (int i = 0; i < points.size(); i++) {
            if (distance(points.get(i).coords, point.coords) <= eps)
                neighbors.add(i);
        }
        return neighbors;
    }

    public static void expandCluster(List<Point> points, int[] clusters, int clusterId,
                                     int pointIdx, List<Integer> neighbors, double eps, int minPts) {
        clusters[pointIdx] = clusterId;
        Queue<Integer> queue = new LinkedList<>(neighbors);

        while (!queue.isEmpty()) {
            int qIdx = queue.poll();
            if (clusters[qIdx] == 0) {
                clusters[qIdx] = clusterId;
                List<Integer> qNeighbors = regionQuery(points, points.get(qIdx), eps);
                if (qNeighbors.size() >= minPts)
                    queue.addAll(qNeighbors);
            }
            if (clusters[qIdx] == -1)
                clusters[qIdx] = clusterId;
        }
    }

    public static Map<Integer, List<Point>> dbscan(List<Point> points, double eps, int minPts) {
        int n = points.size();
        int[] clusters = new int[n];
        int clusterId = 0;

        for (int i = 0; i < n; i++) {
            if (clusters[i] != 0) continue;

            List<Integer> neighbors = regionQuery(points, points.get(i), eps);
            if (neighbors.size() < minPts)
                clusters[i] = -1;
            else {
                clusterId++;
                expandCluster(points, clusters, clusterId, i, neighbors, eps, minPts);
            }
        }

        Map<Integer, List<Point>> clusterMap = new LinkedHashMap<>();
        for (int i = 0; i < n; i++) {
            if (clusters[i] > 0) {
                clusterMap.computeIfAbsent(clusters[i], k -> new ArrayList<>()).add(points.get(i));
            }
        }

        System.out.println("\nResults:");
        int total = points.size();
        for (int id : clusterMap.keySet()) {
            List<Point> cluster = clusterMap.get(id);
            double percent = (cluster.size() * 100.0) / total;
            System.out.printf("Cluster %d: %d points (%.2f%%) → ", id, cluster.size(), percent);
            for (Point p : cluster) System.out.print(p.name + " ");
            System.out.println();
        }

        List<String> noise = new ArrayList<>();
        for (int i = 0; i < n; i++)
            if (clusters[i] == -1)
                noise.add(points.get(i).name);

        if (!noise.isEmpty()) {
            double percent = (noise.size() * 100.0) / total;
            System.out.printf("Noise points: %d (%.2f%%) → %s\n", noise.size(), percent, noise);
        }

        System.out.println("\nCentroids:");
        for (int id : clusterMap.keySet()) {
            double[] mean = centroid(clusterMap.get(id));
            System.out.printf("Centroid %d: [%.2f, %.2f]\n", id, mean[0], mean[1]);
        }

        return clusterMap;
    }

    public static double[] centroid(List<Point> cluster) {
        int dims = cluster.get(0).coords.length;
        double[] mean = new double[dims];
        for (Point p : cluster)
            for (int i = 0; i < dims; i++)
                mean[i] += p.coords[i];
        for (int i = 0; i < dims; i++)
            mean[i] /= cluster.size();
        return mean;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter CSV filename: ");
        String filename = sc.nextLine();

        List<Point> points = readCSV(filename);
        computeDistanceMatrix(points);

        System.out.print("\nEnter epsilon: ");
        double eps = sc.nextDouble();
        System.out.print("Enter minPts: ");
        int minPts = sc.nextInt();

        dbscan(points, eps, minPts);
    }
}
